TST/MAINT: Parametrize `_METRICS_NAMES`

h-vetinari opened the pull request:
This is a follow up to #15054, where I had originally added the following to the deprecation of `wminkowski`:

> While I was add it, I saw a lot of loops over `_METRICS_NAMES` in `test_distance.py` that lent themselves well to parametrization. 

@peterbell10 [commented](https://github.com/scipy/scipy/pull/15054#discussion_r751328277) (before I removed the commits from that PR):
> Can you move the test changes to another PR and consider just using pytest.mark.parametrize instead of a parametrized fixture?

I actually disagree with this one - while I'm not attached to the name, I think this is a prime example of what should be a fixture - every test that wants to test against all distance functions can just use this, and I find it much more ergonomic than adding
```
@pytest.mark.parametrize("metric", _METRICS_NAMES)
```
in front of every test. That said, I could also live with that.

tupui commented on scipy/spatial/conftest.py:
```diff
@@ -0,0 +1,11 @@
+from scipy.spatial.distance import _METRICS_NAMES
+
+
+@pytest.fixture(params=_METRICS_NAMES)
```
We should do this once per session only
```suggestion
@pytest.fixture(params=_METRICS_NAMES, scope="session")
```

tupui commented on scipy/spatial/tests/test_distance.py:
```diff
@@ -398,28 +398,37 @@ def setup_method(self):
+        kwargs = {'N0tV4l1D_p4raM': 3.14, "w": np.arange(3)}
+        args = [3.14] * 200
+
+        assert_raises(TypeError, cdist, X1, X2,
```
Since we are at it, we should match all of these with the message that would be raised.

`with pytest.raises(TypeError, match=r"Message...")`

h-vetinari replied:
I started looking at the error messages here, and the first three distances I tried gave me three different messages (conceptually). Adding a `match=` clause can & should be done in a later PR.

h-vetinari replied:
github seems to have eaten on of your comments about this (did I respond? 🤔), but in any case, the messages were too divergent to try to unify them even with regular expressions - leaving this for a follow-up.

tupui commented on scipy/spatial/tests/test_distance.py:
```diff
@@ -631,31 +637,29 @@ def test_striding(self):
-            x1 = np.random.rand(10, 10)
-            x2 = np.random.rand(10, 10)
+    def test_cdist_refcount(self, metric):
+        x1 = np.random.rand(10, 10)
```
We can also make these deterministic by using a seeded `RandomState`. Also `random` instead of `rand`

h-vetinari replied:
`RandomState` is the [legacy](https://numpy.org/doc/stable/reference/random/legacy.html) API, see [overview](https://numpy.org/doc/stable/reference/random/) and [differences](https://numpy.org/doc/stable/reference/random/new-or-different.html#new-or-different)

The newer `default_rng -> Generator` is available since 1.17, which is I think the oldest numpy we're supporting anyway.

This is also veering quite far away from the intent of this PR, but if you want, I can do that.

h-vetinari replied:
Yup, 1.16 was actually [dropped](https://github.com/scipy/scipy/pull/14655) by you. :)

tupui replied:
I know 😉

We had a few discussions on this when I made the switch. I also wanted to migrate everything to `Generator`, but it is still advices to use `RandomState` for the tests. The reason is that `RandomState` is not going to be removed and will stay there for this purpose while `Generator` is allowed to have its bite stream altered.

h-vetinari replied:
Sure, but these tests are already running in an unseeded manner now. Surely the change from of default from `PCG64` to `PCG64DXSM` is irrelevant here? And if desired, I could already use the latter explicitly.

I just disagree with creating further tech debt if there is no strong reason to do it.

tupui replied:
Here we would reduce the technical debt by using `RandomState.random` not the other way around. It's not advised to use `Generator` there and as I said `RandomState` is kept in NumPy partially for that reason. Here for more information https://numpy.org/neps/nep-0019-rng-policy.html

Appart when we do hypothesis testing, we want to have reproducible tests to ensure we can exactly reproduce failing tests.

h-vetinari replied:
> Here we would reduce the technical debt by using `RandomState.random` not the other way around.

Reducing it sure, but introducing new one (`RandomState` instead of `Generator`)

> It's not advised to use `Generator`

I disagree with that advice.

> we want to have reproducible tests to ensure we can exactly reproduce failing tests.

So reproducibility is good, fully agreed. But I actually very much dislike seeded randomness in tests, because we may just be continually testing a benign/pathological example. Hypothesis is indeed in the sweet spot here - generally random, but with clear seeds upon failure to reproduce it.

However, the tests in this module - as of right now - are not seeded, hence not dependent on such incredible reproducibility that we couldn't survive some very rare changes in the bitstream.

tupui replied:
> > It's not advised to use `Generator`
> 
> I disagree with that advice.

I got burned a few times here 😅 I also wanted at multiple time to go with `Generator` and was advised not to for tests. It's explained in the NEP as well.

That being said, I would personally validate both usage of `Generator` and `RandomState`. It should not really matter and in both cases would be an improvement 😃 

> So reproducibility is good, fully agreed. But I actually very much dislike seeded randomness in tests, because we may just be continually testing a benign/pathological example. Hypothesis is indeed in the sweet spot here - generally random, but with clear seeds upon failure to reproduce it.

It's always a difficult point. We are currently looking into adding some hypothesis testing for some stats, but it's tricky. There is a tradeoff to find because hypothesis testing is costly and what we do should bring value.

> However, the tests in this module - as of right now - are not seeded, hence not dependent on such incredible reproducibility that we couldn't survive some very rare changes in the bitstream.

There is this concern, but in this case it's something else. When there is a test failure which involves random numbers, if we don't have the background on the specific test we always wonder about the source of the issue and if the RNG is somehow involved. I've seen all cases happened and having a seed does help to identify faster the issue.

h-vetinari replied:
> > @h-vetinari: I disagree with that advice.
>
> @tupui: I got burned a few times here 😅 I also wanted at multiple time to go with Generator and was advised not to for tests. It's explained in the NEP as well.

Don't get me wrong, I will go along with what is accepted practice here, but I don't have to agree with it. 😋

I read the NEP, and I think the quote here fits my point well (my emphasis):
> NEP19: The purpose of `RandomState` will be documented as providing certain fixed functionality for _**backwards compatibility**_ and stable numbers for the _**limited purpose**_ of unit testing, and not making whole programs reproducible across numpy versions.

The tests in question here _absolutely do not_ need to be stable with regard to randomness generated in previous numpy versions (indeed, the tests were unseeded up until now).

> @tupui: It's always a difficult point. We are currently looking into adding some hypothesis testing for some stats, but it's tricky. There is a tradeoff to find because hypothesis testing is costly and what we do should bring value.

Yeah, it will be a longterm effort. Still, I think it's good to reflect on why we want to (generally) use seeds - it helps with reproducibility, which helps with debugging, which helps with the scarcest resource: developer time (which matches with what you write in your last point). But _the tests themselves_ should not care about seed they get run with - otherwise there's a bug somewhere (which - however - we don't want to be confronted with intermittently)**.

Not annoying developers with flaky CI and irreproducible failures is a valid and worthwhile goal, locking the tests into a given seed & outcome is not (IMO).

** catching the latent bugs that are sleeping beneath fixed seeds (while still being respectful of developer time) is what hypothesis is very good at, but that'll be a long maturation process of the test suite (including knowing when/where to _not_ use it).

h-vetinari replied:
I think the extent of the discussion here is a good indicator that we shouldn't include these changes in this PR. 🙃

tupui commented on scipy/spatial/tests/test_distance.py:
```diff
@@ -1427,78 +1438,75 @@ def test_pdist_calling_conventions(self):
+        # test that output is numerically equivalent
+        _assert_within_tol(Y_test1, Y_right, eps)
+        # test that Y_test1 and out1 are the same object
+        assert_(Y_test1 is out1)
```
Use plain `assert`

h-vetinari replied:
Plain asserts are also not exactly the height of modern test suites, but it might still be better than `assert_()`

tupui replied:
What do you mean?

h-vetinari replied:
See e.g. https://pypi.org/project/flake8-assert-msg/.

tupui replied:
Thanks I didn't know this hook. This is not directed to tests but when you have assertions in the code itself to explain them. Pytest don't say anything about adding messages and tests should be self explanatory as stated in the doc (they also advise not to write docstrings).

h-vetinari replied:
Fair enough. I guess I just prefer more functional statements like `np.testing.assert_equal` than the statement-syntax of `assert`, but they do have their use of course.

tupui replied:
I agree that it's not very natural also because running `python -O` removes assertions.

tupui reviewed (commented):
Thanks @h-vetinari. Since we are rewriting this, I suggest we take the opportunity to modernize a bit the test suite. I left a few suggestions.

h-vetinari reviewed (commented):
Hey @tupui, thanks for the review! I've followed up on most of your comments, but I want to note that this PR had a pretty narrow focus so far (also with a small diff when ignoring whitespace changes).

Some of the cleanups you're asking for can lead into pretty deep rabbit holes & discussions, and I don't think we should tack on more and more just because this PR is "touching" (really, changing indentation) a bunch of lines here.

h-vetinari commented:
This will also still need a resolution of the issue mentioned in [this](https://github.com/scipy/scipy/issues/14360#issuecomment-974644897) comment.

tupui commented:
> Hey @tupui, thanks for the review! I've followed up on most of your comments, but I want to note that this PR had a pretty narrow focus so far (also with a small diff when ignoring whitespace changes).
> 
> 
> 
> Some of the cleanups you're asking for can lead into pretty deep rabbit holes & discussions, and I don't think we should tack on more and more just because this PR is "touching" (really, changing indentation) a bunch of lines here.

Thanks for the update @h-vetinari.

Regarding the exception, I did not check and would have thought the messages would be similar. If just a few words are changed, `match` support regex and could have been used. If it's too involved then sure we leave this for another PR. I agree that the current PR has a well defined scope and that we should not do too much more.

h-vetinari reviewed (commented):
@tupui, this is ready from my side.

h-vetinari commented:
Side note, I don't know why the test suite timed out on two Azure workers (I haven't added any tests here, or changed how long they take).

I tried restarting them, but got notified that I do not have the rights to do so. This might be due to Azure insisting to link my accounts and getting confused between github and another email address, but just wanted to check if this might be an org-level restriction/decision? CC @rgommers

tupui commented:
> Side note, I don't know why the test suite timed out on two Azure workers (I haven't added any tests here, or changed how long they take).
> 
> I tried restarting them, but got notified that I do not have the rights to do so. This might be due to Azure insisting to link my accounts and getting confused between github and another email address, but just wanted to check if this might be an org-level restriction/decision? CC @rgommers

It happens from time to time. This is just some Azure not being 100% deterministic with the performances it provides. As for restarting jobs, I do not have access either. I don't know how this is handled.

rgommers commented:
IIRC there's a SciPy team on Azure, which requires having an account. And then what happens for me is that if I follow links from here to a failing job, I'm typically not logged in; it requires a lot of clicks. Anyway, please don't bother restarting failing jobs unless there's a reason why that failing job is special. If all the other Azure jobs pass, it's fine to ignore.

h-vetinari commented:
> Anyway, please don't bother restarting failing jobs unless there's a reason why that failing job is special. If all the other Azure jobs pass, it's fine to ignore.

OK thanks - I just suspected that it has something to do with my pretty busted azure account setup (since microsoft forcefully merges everything it ever encounters in the same context, leaving a mess).

I'm also used to a different workflow in conda-forge, where jobs often need to be restarted, and it sometimes is actually the most effective. Finally, conda-forge allows me to retrigger azure jobs from the github CI (not sure how that's set up).

In any case, thanks for the clarifications! :)

h-vetinari commented:
Rebased after https://github.com/scipy/scipy/pull/15055 went in

tupui commented:
@h-vetinari I might be short on time for this one. But there is no rush for 1.9 since these are just tests related changes.

peterbell10 commented on scipy/spatial/tests/conftest.py:
```diff
@@ -0,0 +1,11 @@
+
+
+@pytest.fixture(params=_METRICS_NAMES, scope="session")
+def metric(request):
```
Can't this be defined in `test_distance.py` since it's the only file that needs it? That might help with the meson failures as well.

h-vetinari replied:
That fixture should be defined wherever `_METRICS_NAMES` is defined. I can move the whole thing to `test_distance.py` though.

h-vetinari replied:
Moved to `test_distance.py` now.

h-vetinari commented:
Hello @tupui @peterbell10, I've dusted off this old PR and cleaned it up. There's no extra `conftest.py` anymore (and I've backed out the `assert_` changes as well), so it should be pretty simple & uncontroversial now I hope. :)

I've rebased the commits in such a way that the one with the all the indentation changes can be easily reviewed separately (using ⚙️ in the UI to hide whitespace changes), then the rest is straight-forward.

h-vetinari commented:
@peterbell10, could you let us know if you're still interested in reviewing here? Otherwise I'd go ahead and merge this in ~48h.
